components/Datepicker/src/index.tsx

```diff
@@ -0,0 +1,188 @@
+import React from 'react';
+import clsx from 'clsx';
+import {
+  format,
+  Locale,
+  getWeeksInMonth,
+  formatISO,
+  setDay,
+  addWeeks,
+  subMonths,
+  addMonths,
+  isSameMonth,
+  setDate,
+} from 'date-fns';
+import { nl } from './locale';
+
+import BaseProps from '@gemeente-denhaag/baseprops';
+
+import './datepicker.css';
+
+export interface DatepickerProps extends BaseProps {
+  /**
+   * Whether the input field is in a error state.
+   */
+  error?: boolean;
+  /**
+   * Whether the input field is disabled.
+   */
+  disabled?: boolean;
+  /**
+   * Locale (from @gemeente-denhaag/datepicker/locale or date-fns/locale) (default: nl)
+   */
+  locale?: Locale;
+  /**
+   * Starting date for calendar (default: today)
+   */
+  startDate?: Date;
+}
+
+interface DatepickerState {
+  current: Date;
+  selected?: Date;
+  opened: boolean;
+}
+
+/**
+ * The DatePicker component allows the user to select a date.
+ */
+export const Datepicker: React.FC<DatepickerProps> = ({
+  startDate = new Date(),
+  locale = nl,
+  ...props
+}: DatepickerProps) => {
+  const rootStyles = clsx(
+    'denhaag-datepicker',
+    {
+      'denhaag-datepicker--error': props.error,
+      'denhaag-datepicker--disabled': props.disabled,
+    },
+    props.className,
+  );
+
+  const [state, setState] = React.useState<DatepickerState>({
+    current: setDate(startDate, 1),
+    selected: undefined,
+    opened: false,
+  });
+
+  const componentRef = React.useRef<HTMLDivElement>(null);
+
+  const outsideClickListener = (e: Event) => {
+    if (!componentRef.current?.contains(e.target as Node)) {
+      setState({
+        ...state,
+        opened: false,
+      });
+    }
+  };
+
+  React.useEffect(() => {
+    document.addEventListener('click', outsideClickListener);
+    return () => {
+      document.removeEventListener('click', outsideClickListener);
+    };
+  }, [componentRef]);
+
+  const renderDay = (i: number, j: number) => {
+    const date = setDay(addWeeks(state.current, i), (j + (locale.options?.weekStartsOn || 0)) % 7, { locale: locale });
+    if (isSameMonth(date, state.current)) {
+      return (
+        <button
+          type="button"
+          onClick={() => {
+            setState({ ...state, selected: date, opened: false });
+          }}
+        >
+          {date.getDate()}
+        </button>
+      );
+    }
+    return '';
+  };
+
+  return (
+    <div className={rootStyles} ref={componentRef}>
```

Comment:
Shouldn't the ref for this component be on the `<input>`? Unless I'm misunderstanding refs.

Reply:
The `componentRef` is used for the `outsideClickListener` so this does have to be on the outer `div` instead of the `input` element. Made a few changes though, exposed `ref` so the user can also add a `ref` of their own (replacing `componentRef`) and added a `inputRef` prop which the user can use to supply a `ref` to the HTML `input` element.